### DISCUSSION TITLE
A cpack package generator to generate vcpkg ports?

### DISCUSSION THREAD

**Discussion by sketch34** [microsoft/vcpkg#43020] [Q&A] [answered]
Is anyone aware of any work being done towards this, or is it on the roadmap?

Something like this:
```
include(CPack)
include(CPackVcpkg)
...
```
```
cpack -G "vcpkg" ...
```

Which produces a canonical vcpkg port (`portfile.cmake`, `vcpkg.json` and additional cmake files to support find_package() etc.) based on the install rules you have defined in your project's cmake.

**Comment by dg0yt** [comment]
TBH I don't see a good match. CPack is for generating installable packages of binary artifacts, but ports are build recipes. The install rules provide little information about features and dependencies. In fact features and portfile change which install rules are activated at all.

**Reply by sketch34** [reply]
You might be right. But the info available to the "External" generator suggests that there could be sufficient info available to CPack to produce a vcpkg port?

> The External generator was added in CMake 3.13 and is very different to all the other package
generators. This particular generator writes out a JSON file containing package metadata,
component details, and other CPack information, but it does not produce a package itself. It can
install the files that would normally be packaged into a temporary staging area if requested. Other
tools are expected to consume the JSON file and staged install area to produce packages using their
own methods. The main goal of this generator is to present the details accumulated by CPack in a
way that platforms and distributions can easily use within their own policy and technical
constraints.
>
> Professional CMake: A Practical Guide 19th Edition

> The JSON metadata file contains a list of CPack components and component groups, the various options passed to [cpack_add_component()](https://cmake.org/cmake/help/latest/module/CPackComponent.html#command:cpack_add_component) and [cpack_add_component_group()](https://cmake.org/cmake/help/latest/module/CPackComponent.html#command:cpack_add_component_group), the dependencies between the components and component groups, and various other options passed to CPack.
>
> https://cmake.org/cmake/help/latest/cpack_gen/external.html

But as you say perhaps this is too far removed from the intention of CPack and there is still insufficient info / not a sensible mapping from CPack -> vcpkg port.

**Reply by dg0yt** [reply, marked as answer]
AFAIU this is the information which provides the installated files, associated with *installation* components, of one particular build configuration. vcpkg always collects all installed files in one binary packages.

The information doesn't provide dependencies, and it is blind against any input variables (or dependencies) which enable optional features. But this is the key information for manifest (dependencies, features) and portfile (wiring feaurees and dependencies to project configuration).
